### PR TITLE
First commit genform

### PR DESCRIPTION
**Pull request by chufz**
addition of genform tool, see @bernt-matthias

### CODE REVIEW

**Comment by bernt-matthias** [comment]
@fgiacomoni could you give an example of what Infos you are missing?

I also thought that a bit more structure would be nice for the tool. But it seems quite a bit of work.

File: tools/genform/.shed.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+categories: Metabolomics
```
[the remark this answers is not in the document]
Comment (lecorguille) [reply tools/genform/.shed.yml]: It's either 
```
categogies: [Metabolomics]
```
or 
```
categogies:
- Metabolomics
```

File: tools/genform/genform.xml [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+    tail -n \$hwc > '$out'
+  #else
+    && cp out '$out'
+  #end if
+    ]]></command>
+    <inputs>
+        <param type="data" argument="ms" label="MS data" format="tabular" help="" />
+        <param type="data" argument="msms" label="MS/MS data" optional="true" format="tabular" help="" />
+        <conditional name="formula_cond">
+            <param type="select" name="formula_select" label="structural formula existence check" help="allow only molecular formulas for that at least one structural formula exists or specify advanced restictions, note that the default of genform is 'no existence check'">
```
[the remark this answers is not in the document]
Comment (lecorguille) [reply tools/genform/genform.xml]: @bernt-matthias 
```xml
        <conditional name="elemfilt_cond">
            <param type="select" name="elemfilt_select" optional="true" label="chemical element filtering">
                <option value="elem">by used elements</option>
                <option value="oc">only organic compounds (oc)</option>
                <option value="fuzzy">by fuzzy formula</option>
            </param>
```
Galaxy will add an option `Nothing selected` at the top of your list

**Comment by lecorguille** [comment]
@fgiacomoni 
Do you need more info than: 
https://github.com/workflow4metabolomics/tools-metabolomics/blob/cdf23abb87182b74f2009413d8c48c1c77f33107/tools/genform/.shed.yml

File: tools/genform/genform.xml [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+Example output:
+
+======== === =========== ====== ======== ========= ========
+C4H9N3O2  2   132.07675   -0.8   96.674   100.000   96.674
+======== === =========== ====== ======== ========= ========
+
+**Advanced options**
+
+...
```
Comment (lecorguille) [inline tools/genform/genform.xml]: Can be frustrating :)

Comment (chufz) [reply tools/genform/genform.xml]: I know, will change that...
is a link to the command line tool manual sufficient enough?
[https://www.researchgate.net/publication/307964728_MOLGEN-MSMS_Software_User_Manual](url)

Comment (lecorguille) [reply tools/genform/genform.xml]: It's ok for me

Comment (fgiacomoni) [reply tools/genform/genform.xml]: it's ok for me too. Yes the information available in the yaml file is enough but it is not evident for anybody to find the information in this file (more in a README ou markdown file) ;-)

Comment (fgiacomoni) [reply tools/genform/genform.xml]: Do you know the HR2 algorithm developped by Tobias Kind and Oliver Fiehn. (2007) ? Publication is I think "Seven Golden Rules for heuristic filtering of molecular formulas obtained by accurate mass spectrometry." BMC Bioinformatics p8:105 http://www.ncbi.nlm.nih.gov/pubmed/17389044
HR2 original program: http://fiehnlab.ucdavis.edu/projects/Seven_Golden_Rules/Software

**Comment by lecorguille** [comment]
So where are we here?

File: tools/genform/genform.xml (hunk 3 above)
Comment (chufz) [reply tools/genform/genform.xml]: Ok so I add a link to the commandline manual

**Comment by chufz** [comment]
> So where are we here?

So far, I think all requested changes should be resolved, @bernt-matthias &me will have time for it end of next week to look at if you find more fixes needed.

File: tools/genform/genform.xml [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+genform 
+  ms='$ms'
+  #if $msms
+    msms='$msms'
+  #end if
+  #if $formula_cond.formula_select != 'advanced'
+    $formula_cond.formula_select
+  #else
+    $formula_cond.vsp
+    #if str($formula_cond.vsm2mv) != ''
```
Comment (bgruening) [inline tools/genform/genform.xml]: `#if str($formula_cond.vsm2mv):` should be enough

File: tools/genform/genform.xml [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+  #else
+    && cp out '$out'
+  #end if
+    ]]></command>
+    <inputs>
+        <param type="data" argument="ms" label="MS data" format="tabular" help="" />
+        <param type="data" argument="msms" label="MS/MS data (optional)" optional="true" format="tabular" help="" />
+        <conditional name="formula_cond">
+            <param type="select" name="formula_select" label="Structural formula existence check" help="Allow only molecular formulas for that at least one structural formula exists or specify advanced restictions, note that the default of genform is 'no existence check'">
+                <option value="">No existence check</option>
```
Comment (bgruening) [inline tools/genform/genform.xml]: you can remove this and make it optional=true?

File: tools/genform/genform.xml [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+            <option value="msmv=nsse">Normalized sum of squared</option>
+            <option value="msmv=nsae" selected="true">Absolute errors</option>
+        </param>
+        <param type="integer" argument="acc" label="Allowed deviation for full acceptance" min="0" value="2" help="of MS/MS peak in ppm" />
+        <param type="integer" argument="rej" label="Allowed deviation for total rejection" min="0" value="4" help="of MS/MS peak in ppm" />
+        <param type="float" argument="thms" label="Threshold for the MS match value" min="0" max="1" value="0" help="" />
+        <param type="float" argument="thmsms" label="Threshold for the MS/MS match value" min="0" max="1" value="0" help="" />
+        <param type="float" argument="thcomb" label="Threshold for the combined match value" min="0" max="1" value="0" help="" />
+        <param type="select" argument="sort" label="Sort generated formulas by" help="" >
+            <option value="">Do not sort</option>
```
Comment (bgruening) [inline tools/genform/genform.xml]: optional=true instead?

File: tools/genform/genform.xml [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+            <option value="sort=ppm">Mass deviation in ppm</option>
+            <option value="sort=msmv">MS match value</option>
+            <option value="sort=msmsmv">MS/MS match value</option>
+            <option value="sort=combmv">Combined match value</option>
+        </param>
+        <conditional name="elemfilt_cond">
+            <param type="select" name="elemfilt_select" label="Chemical element filtering">
+                <option value="nofilter">No filter</option>
```
Comment (bgruening) [inline tools/genform/genform.xml]: optional=true

**Comment by bernt-matthias** [comment]
Hey @bgruening . Most of your comments should now be covered. Only: I can't make the selects within conditional (which would actually be neat: https://github.com/galaxyproject/galaxy/issues/7173)

**Comment by lecorguille** [comment]
It seems that there is an anti-spam system in researchgate:

``` .. ERROR: Error '429 Client Error: Too Many Requests for url: https://www.researchgate.net/publication/307964728_MOLGEN-MSMS_Software_User_Manual' accessing https://www.researchgate.net/publication/307964728_MOLGEN-MSMS_Software_User_Manual
```

**Comment by bgruening** [comment]
Let's download this PDF and put it on cargo-port.

File: tools/genform/.shed.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/genform/.shed.yml]: done

**Comment by bernt-matthias** [comment]
Found a newer version of the manual on citeseer .. but deposited it on cargoport anyway: https://github.com/galaxyproject/cargo-port/pull/163 ... will fix the link here once this is merged

**Comment by lecorguille** [comment]
Cool!

**Comment by lecorguille** [comment]
Hop: [toolshed.g2.bx.psu.edu](https://toolshed.g2.bx.psu.edu/view/workflow4metabolomics/genform/963722c4e1da)

Thanks @chufz @bernt-matthias for this new tool within the metabolomics community

Thanks @bgruening for your review

**Pull request merged**